### PR TITLE
Added std capture and it's corresponding flag

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Very good jobs, requesting a few tweaks in comments

File: protostar/commands/test/test_cases.py [hunk 1]
```diff
@@ -22,6 +22,7 @@ def get_formatted_file_path(self):
 class PassedTestCase(TestCaseResult):
     test_case_name: str
     execution_resources: Optional[ExecutionResourcesSummary]
+    logs: str
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/test_cases.py:25]: And as @kasperski95 we have logger, it is name collision

File: protostar/commands/test/test_cases.py [hunk 2]
```diff
@@ -17,13 +19,19 @@ class TestCaseResult:
...
 
 @dataclass(frozen=True)
 class PassedTestCase(TestCaseResult):
     test_case_name: str
     execution_resources: Optional[ExecutionResourcesSummary]
+    captured_setup_stdout: str
+    captured_test_stdout: str
 
-    def __str__(self) -> str:
+    def log(self, include_stdout_section: bool = False) -> str:
```
Comment (mkaput) [inline protostar/commands/test/test_cases.py:34]: maybe?

```suggestion
    def display(self, include_stdout_section: bool = False) -> str:
```

by the name `log` I thought I can send custom logs from testing code with this function

File: protostar/commands/test/test_cases.py [hunk 3]
```diff
@@ -75,14 +105,36 @@ def __str__(self) -> str:
...
+                    "[setup]:\n"
+                    f"{log_color_provider.colorize('GRAY', self.captured_setup_stdout)}\n"
+                )
+
+            if self.captured_test_stdout:
+                result.append(
+                    "[test]:\n"
+                    f"{log_color_provider.colorize('GRAY', self.captured_test_stdout)}\n"
+                )
+
```
Comment (mkaput) [inline protostar/commands/test/test_cases.py:137]: this section is repeated in PassedTestCase

File: protostar/commands/test/test_cases.py [hunk 4]
```diff
@@ -17,13 +19,19 @@ class TestCaseResult:
     def get_formatted_file_path(self):
         return log_color_provider.colorize("GRAY", str(self.file_path))
 
+    @abstractmethod
+    def log(self, _: bool = False) -> str:
```
Comment (mkaput) [inline protostar/commands/test/test_cases.py:23]:
```suggestion
    def log(self, include_stdout_section: bool = False) -> str:
```

and please use the same argument name in each override.

you are breaking `kwargs` passing here. parameter names are parts of function contract in Python, following snippet would crash here:

```python
result = get_random_test_case_result()
result.log(include_stdout_section=True)
# or
result.log(_=True)
```

File: website/docs/cli-reference.md [hunk 5]
```diff
@@ -110,14 +110,16 @@ Additional directories to look for sources.
 Disable hint validation in contracts declared by the `declare` cheatcode or deployed by `deploy_contract` cheatcode.
 
 #### `-x` `--exit-first`
-Exit instantly on first broken or failed test
+Exit immediately on first broken or failed test
```
Comment (mkaput) [inline website/docs/cli-reference.md:113]: could you please move this change to separate PR?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/commands/test/test_cases.py [hunk 6]
```diff
@@ -17,13 +19,47 @@ class TestCaseResult:
...
 
+    @abstractmethod
+    def display(self, include_stdout_section: bool = False) -> str:
+        ...
+
+
+@dataclass(frozen=True)
+class FinishedTestCase(TestCaseResult):
+    captured_setup_stdout: str
+    captured_test_stdout: str
```
Comment (mkaput) [inline protostar/commands/test/test_cases.py:30]: I think that it might be nice to have this model generalized over a possibility of having 0:* stdouts captured in test case. This way, it would be natural to write generic code over `setup` and `test` stages.

I'm thinking about following data model:

```python
class PassedTestCase:
    stdouts: List[OutputCapture]

@dataclass
class OutputCapture
    # typing pseudocode for brevity
    stream: "stdout" | "stderr" # possible future addition
    stage: str # "setup" or "test" but can be arbitrary
    data: bytes
```

It is also trivial to express interleaving of stdout/stderr ✖️ setup/test/whatever captures here. You just insert `OutputCapture`s for each captured line, one by one as encountered` to the list.

Comment (mkaput) [reply protostar/commands/test/test_cases.py:30]: I do wonder if we actually have to capture `setup` stdout ? I haven't ever needed such feature

Comment (mkaput) [reply protostar/commands/test/test_cases.py:30]: discussed privately, let's leave current implementation and consider this refactoring if we stumble upon a need for it

File: protostar/commands/test/test_cases.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/test_cases.py]: Problem is we use `log` a lot as a name for such functions, I am worried about consistency. Agreed `display` is a better name though.

Comment (MaksymilianDemitraszek) [reply protostar/commands/test/test_cases.py]: Oh i didn't see nvm then

File: protostar/commands/test/test_cases.py [hunk 7]
```diff
@@ -7,27 +8,53 @@
 )
 from protostar.commands.test.test_environment_exceptions import ReportedException
 from protostar.protostar_exception import UNEXPECTED_PROTOSTAR_ERROR_MSG
-from protostar.utils.log_color_provider import log_color_provider
+from protostar.utils.log_color_provider import log_color_provider, SupportedColorName
+
+
+def _get_formatted_stdout(
```
Comment (mkaput) [inline protostar/commands/test/test_cases.py:14]: nitpick: could you please move these underscored functions to bottom of the file, so that the meat of this module stays on top? :)

File: protostar/commands/test/test_command.py [hunk 8]
```diff
@@ -122,6 +128,7 @@ async def test(
         no_progress_bar=False,
         fast_collecting=False,
         exit_first=False,
+        stdout_on_success=False,
```
Comment (mkaput) [inline protostar/commands/test/test_command.py:131]:
```suggestion
        stdout_on_success: bool = False,
```

File: protostar/commands/test/test_command.py [hunk 9]
```diff
@@ -97,6 +97,11 @@ def arguments(self) -> List[Command.Argument]:
                 type="bool",
                 description="Exit immediately on first broken or failed test",
             ),
+            Command.Argument(
+                name="stdout-on-success",
+                type="bool",
+                description="Also print captured stdout for passed test cases.",
```
Comment (mkaput) [inline protostar/commands/test/test_command.py:103]:
```suggestion
                description="Print captured standard output for passed test cases.",
```

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/commands/test/test_command.py (hunk 9 above)
Comment (Radinyn) [reply protostar/commands/test/test_command.py:103]: This "also" is here to show that the default behaviour is that we always print for failed test cases

File: protostar/commands/test/test_command.py (hunk 8 above)
Comment (Radinyn) [reply protostar/commands/test/test_command.py:131]: We don't do that anywhere else, is it necessary here?

**Review by mkaput — APPROVED** [review]
